### ISSUE TITLE
Generate small number of very, very long vectors

### ISSUE THREAD

**Issue by kgoldfeld** [mfasiolo/mvnfast#14]
This is a great package - I've relied on it for my [simstudy](https://kgoldfeld.github.io/simstudy/dev/index.html) package. I am simulating data from a scenario where I need to generate correlated data for a very large cluster size - say 10000+.  Typically, the number of clusters will be small (<100). However, even mvnfast breaks down. Do you have any suggestions about how I might approach this problem, or in your opinion, is it just not possible? Are there memory adjustments I can make?

**Comment by mfasiolo** [comment]
Hi, do you have a minimal reproducible example of what you would like to do with mvnfast?

**Comment by kgoldfeld** [comment]
Sorry about that. I played around a little more, and it looks like the Cholesky decomposition is the issue. Below is an example with a target vector of length 1000, however, I need to go up to 10000 or 20000, in which case it seems impossible:

First, the variance-covariance matrix:

```
library(microbenchmark)
library(mvnfast)

n <- 1000
rho <- 0.33
Cov <- matrix(rho, nrow = n, ncol = n)
diag(Cov) <- 1
```
The data generation with the `mvnfast` package:
  
```
microbenchmark(rmvn(n = 1, mu = rep(0, n) , sigma = Cov))
#> Unit: milliseconds
#>                                      expr      min       lq     mean   median
#>  rmvn(n = 1, mu = rep(0, n), sigma = Cov) 160.7192 161.3706 161.9238 161.5539
#>        uq      max neval
#>  162.1066 166.9593   100
```

I can speed things up a tad using the `Rfast` package and my own crude function:
  
 ```
genMN <- function(Cov) {
    n <- nrow(Cov)
    L <- t(Rfast::cholesky(Cov))
    Z <- rnorm(n)
    as.vector(L %*% Z)
  }

microbenchmark(genMN(Cov))
#> Unit: milliseconds
#>        expr      min       lq     mean   median       uq      max neval
#>  genMN(Cov) 130.8132 131.5412 134.2275 132.9951 133.8848 193.4751   100
```

But, `Rfast`  only improves things so much. Once we hit 5000, it just takes too long. Here is a comparison with 3 different options (still at 1000). The time for the data generation is almost entirely the decomposition, but you know that already.
  
```
microbenchmark(
    base::chol(Cov),
    Matrix::chol(Cov),
    Rfast::cholesky(Cov)
  )
#> Unit: milliseconds
#>                  expr      min       lq     mean   median       uq      max
#>       base::chol(Cov) 159.0344 160.0882 162.3748 160.8902 162.2870 218.6505
#>     Matrix::chol(Cov) 159.4029 160.0869 161.5528 160.8012 162.0993 174.3313
#>  Rfast::cholesky(Cov) 128.4068 129.2809 130.4548 129.6620 130.9135 148.4977
#>  neval cld
#>    100  a 
#>    100  a 
#>    100   b
```

So, we need a faster decomposition, which I am assuming is not possible.

**Comment by mfasiolo** [comment]
I agree with your conclusion, you could speed up things by parallelizing the Cholesky decomposition (but I am not aware of parallel Cholesky decompositions callable from R) or by exploiting the structure of the covariance matrix (e.g., in cases where you are dealing with sparse covariance matrices).
